added junitreport tool

stevekuznetsov opened the pull request:
# junitreport

`junitreport` is a tool that allows for the consumption of test output in order to create jUnit XML.

## Installation

In order to build and install `junitreport`, from the root of the OpenShift Origin repository, run `hack/buil-go.sh cmd/junitreport`. 

## Usage 

`junitreport` can read the output of different types of tests. Specify which output is being read with `--type=<type>`. Supported test output types currently include `'gotest'`, for `go test` output. The default test type is `'gotest'`. 

`junitreport` can output flat or nested test suites. To choose which type of output to use, set `--suites=<type>` to either `'flat'` or `'nested'`. The default suite output structure is `'flat'`. When creating nested test suites, `junitreport` will use `/` as the delimeter between suite names: `github.com/maintainer/repository/suite` will be parsed as a hierarchy of `github.com`, `github.com/maintainer`, *etc.* If you are requesting nested test suite output but do not want the root suite(s) to be as general as `github.com`, for example, set `--roots=<root suite names>` to be a comma-delimited list of the names of the suites you wish to use as roots. If the parser encounters a package outside of those roots, it will ignore it. This allows a user to provide a root suite and only collect data for children of that root from a larger data set.

Ensure that the output you are feeding `junitreport` is free of extraneous text - any lines that are not test/suite declarations, metadata, or results are interpreted as test output. Text that you do not expect to see in Jenkins, for example, while looking at the output of a failed test should not be included in the input to `junitreport`.

### Examples

To parse the output of `go test` into a flat collection of test suites:

```sh

$ go test -v -cover ./... | junitreport > report.xml
```

To parse the output of `go test` into a nested collection of test suites rooted at `github.com/maintainer`:

```sh

$ go test -v -cover ./... | junitreport --suites=nested --roots=github.com/maintainer > report.xml
```

stevekuznetsov commented:
Still left to do: 
 * unit testing for `pkg/parser/gotest`
 * unit testing for `pkg/builder`
 * unit testing for `pkg/api`
 * unit testing for `pkg/parser/stack/parser.go`

stevekuznetsov commented:
@deads2k @kargakis PTAL

stevekuznetsov commented on cmd/junitreport/pkg/api/util/time.go:
```diff
@@ -0,0 +1,22 @@
+// XML marshalling functions, so when a time.Duration is marshalled into XML, it is
+// represented as the unit value of nanoseconds in the duration. This ugly hack is
+// the only way to make the marshalled value useable.
+type Duration time.Duration
```
@liggitt  @deads2k Very unhappy about this. If anyone knows of a better way, I'm happy to use it. I do *not* want to store times as strings internally.

liggitt replied:
how does jenkins interpret durations?

stevekuznetsov replied:
`// TODO(skuznets): figure that out`

stevekuznetsov replied:
```xml
               <xs:attribute name="time" type="xs:decimal" use="required">
                  <xs:annotation>
                     <xs:documentation xml:lang="en">Time taken (in seconds) to execute the test</xs:documentation>
                  </xs:annotation>
               </xs:attribute>
```
from http://help.catchsoftware.com/display/ET/JUnit+Format

stevekuznetsov replied:
Changed the field to be a `float64` representing seconds.

liggitt commented on cmd/junitreport/README.md:
```diff
@@ -0,0 +1,31 @@
+
+`junitreport` can read the output of different types of tests. Specify which output is being read with `--type=<type>`. Supported test output types currently include `'gotest'`, for `go test` output. The default test type is `'gotest'`. 
+
+`junitreport` can output flat or nested test suites. To choose which type of output to use, set `--suites=<type>` to either `'flat'` or `'nested'`. The default suite output structure is `'flat'`. When creating nested test suites, `junitreport` will use `/` as the delimeter between suite names: `github.com/maintainer/repository/suite` will be parsed as a hierarchy of `github.com`, `github.com/main ...
```
not sure I'd expect a failure for tests outside specified roots, more that the results would be ignored...

stevekuznetsov replied:
That does sound like better behavior for a given root. I'll see how much complexity it adds to the parser.

liggitt commented on cmd/junitreport/README.md:
```diff
@@ -0,0 +1,31 @@
+
+`junitreport` can output flat or nested test suites. To choose which type of output to use, set `--suites=<type>` to either `'flat'` or `'nested'`. The default suite output structure is `'flat'`. When creating nested test suites, `junitreport` will use `/` as the delimeter between suite names: `github.com/maintainer/repository/suite` will be parsed as a hierarchy of `github.com`, `github.com/main ...
+
+Ensure that the output you are feeding `junitreport` is free of extraneous text - any lines that are not test/suite declarations, metadata, or results are interpreted as test output.
```
details on what is considered extraneous text would be helpful

stevekuznetsov replied:
Anything that isn't actually test output and would be inappropriate to see in Jenkins when looking at the output of a failed test. I'll add that in.

stevekuznetsov commented:
@liggitt Updated behavior and README to ignore test suites that occur outside of the given root suites when using nested suites - this allows for the collection of only a specific sub-tree or sub-trees from a larger set of data.

liggitt commented on cmd/junitreport/pkg/api/string.go:
```diff
@@ -0,0 +1,35 @@
+		message = t.FailureOutput.Message
+		output = t.FailureOutput.Output
+	}
+	result = "passed"
```
result is always "passed"?

stevekuznetsov replied:
It's set differently if the failure or skip messages are present.

kargakis replied:
But even if they are present, won't result still be passed in the end?

stevekuznetsov replied:
Yes, d'oh.

liggitt commented on cmd/junitreport/pkg/api/test_case.go:
```diff
@@ -0,0 +1,38 @@
+	"strings"
+	"time"
+)
+
```
weird file divisions... why isn't the TestCase type and String impl in this file?

stevekuznetsov replied:
Because I don't really know what I'm doing and these started out without methods like in Origin. Should I just collapse it all into one file?

soltysh replied:
Following origin conventions, I'd say package `api` should contain only types. The implementation should be placed somewhere else.

stevekuznetsov replied:
@soltysh @liggitt Are we saying that this needs to change so no serialized objects have methods? We will just have a package of mutating methods for the current contents of `string.go`, *etc*?

liggitt commented on cmd/junitreport/pkg/builder/nested/test_suites_builder.go:
```diff
@@ -0,0 +1,158 @@
+}
+
+const (
+	// TestSuiteNameDelimeter is the default delimeter for test suite names
```
delimiter

stevekuznetsov replied:
No, this is the sandwich meat measuring stick.

stevekuznetsov commented on cmd/junitreport/pkg/api/test_case.go:
```diff
@@ -0,0 +1,38 @@
+	"time"
+)
+
+// SetDuration sets the runtime duration of the test case
```
Because I don't really know what I'm doing and these started out without methods like in Origin. Should I just collapse it all into one file?

liggitt commented on cmd/junitreport/pkg/builder/nested/test_suites_builder.go:
```diff
@@ -0,0 +1,158 @@
+	if !b.isChildOfRoots(name) {
+		return nil, errors.NewSuiteOutOfBoundsError(name)
+	}
+	parentName, exists := b.getParentName(name)
```
`s/exists/hasParent/`, I got confused thinking this was looking up the parent

liggitt replied:
or even just go off of len(parentName) == 0?

liggitt replied:
feels similar to https://golang.org/pkg/path/filepath/#Ext, just return "" for suites without parents

liggitt commented on cmd/junitreport/pkg/builder/nested/test_suites_builder.go:
```diff
@@ -0,0 +1,158 @@
+}
+
+func (b *nestedTestSuitesBuilder) isChildOfRoots(name string) bool {
+	for _, rootSuite := range b.testSuites.Suites {
```
how does this handle cases where you want to consume everything (aren't specifying `--roots`)?

liggitt commented on cmd/junitreport/pkg/builder/nested/test_suites_builder.go:
```diff
@@ -0,0 +1,158 @@
+	}
+	parent.Children = append(parent.Children, suite)
+
+	b.updateMetrics(suite)
```
so suites are assumed to have completed prior to being added?

stevekuznetsov replied:
Yes, `AddSuite` should be called when a completed suite is ready to be added.

liggitt commented on cmd/junitreport/pkg/parser/interfaces.go:
```diff
@@ -0,0 +1,61 @@
+// contains useful information or extract the useful information from it
+type TestDataParser interface {
+	// ContainsIdentifier determines if the line contains the identifier for a test
+	ContainsIdentifier(line string) bool
```
too generic, `ContainsTestIdentifier`?

liggitt replied:
definitely be more specific with go's duck typing.... don't want to accidentally satisfy the wrong interface

liggitt commented on cmd/junitreport/pkg/parser/interfaces.go:
```diff
@@ -0,0 +1,61 @@
+	// ExtractName extracts the name of the test case from test output lines and returns
+	// the name if all of the extracted names match. This extraction supports multiple lines
+	// as the test name often appears at the declaration and the test result line.
+	ExtractName(identifiers []string) (name string, ok bool)
```
`ExtractTestName`? how is `ContainsIdentifier` different than `_, containsIdentifier := ExtractName()`?

liggitt commented on cmd/junitreport/pkg/parser/interfaces.go:
```diff
@@ -0,0 +1,61 @@
+
+// TestDataParser knows how to take raw test data and either characterize it as data that
+// contains useful information or extract the useful information from it
+type TestDataParser interface {
```
not sure the `ContainsX`, `ExtractX` pattern is useful... doesn't that mean you're doing work twice, and would I ever check if a line contains something I wouldn't immediately want to extract?

stevekuznetsov replied:
For instance, tests are identified twice: at the declaration and the result. Collect the data, but don't extract until you have both and can compare to make sure nothing went wrong.

stevekuznetsov replied:
If we want a single interface for extraction we have already agreed that we need `ExtractIdentifer`, `ExtractResult`, and not `identifier, result, whatever, time := ParseGoTestResultLine`. Therefore, we're already multiplying the number of regex calls we are doing. I am not a huge fan of methods that return two values, one of which you need to check to see if the other is useful... so I don't really mind doubling the work again. I agree the current implementation of "figure out if it goes, then stick it in something and extract later" isn't the best, and I'm about to refactor it, but I am not convinced that removing the `ContainsX` methods is a good idea for readability.

stevekuznetsov replied:
We also can't mark a test failed with an output message until we've a) parsed the test result line and b) collected all of the test output.

liggitt commented on cmd/junitreport/pkg/parser/stack/parser.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+// Parse parses output syntax of a specific class, the assumptions of which are outlined in the struct definition.
+// The specific boundary markers and metadata encoding is free to vary as long as it can be expressed as
```
finish the godoc thought... I'm on the edge of my seat

liggitt commented on cmd/junitreport/pkg/parser/stack/parser.go:
```diff
@@ -0,0 +1,95 @@
+// testOutputParser uses a stack to parse test output. Critical assumptions that this parser makes are:
+//   1 - packages may be nested but tests may not
+//   2 - no package declarations will occur within the boundaries of a test
+//   3 - all tests and packages are fully bounded by a start and result line
```
why do packages need both a start and end? you can collect tests into an anonymous package and identify it at the end with the package results, right?

stevekuznetsov replied:
We want to nest packages. Packages without boundaries are difficult to nest. We have complete control over how we present output from our tests, so I didn't see why we would choose to make our lives harder than they need to be.

liggitt commented on cmd/junitreport/pkg/parser/stack/parser.go:
```diff
@@ -0,0 +1,95 @@
+//       the sub-package's or member test's result line will occur before that of the parent package
+//       i.e. any test or package overlap will necessarily mean that one package's lines are a superset
+//       of any lines of tests or other packages overlapping with it
+//   5 - any text in the input file that doesn't match the above regex is necessarily the output of the
```
what regex?

liggitt commented on cmd/junitreport/pkg/parser/stack/parser.go:
```diff
@@ -0,0 +1,95 @@
+
+	for input.Scan() {
+		line := input.Text()
+
```
with this many fallthroughs, a switch is actually harder to follow, sequential ifs would probably be easier

stevekuznetsov replied:
The reason for that is the final default looks really ugly with sequential if. However if like you say we put everything in output then sequential out if the way to go.

liggitt commented on cmd/junitreport/pkg/parser/stack/parser.go:
```diff
@@ -0,0 +1,95 @@
+			currentTestData.Output = append(currentTestData.Output, line)
+		}
+
+		if currentTestData.IsComplete() {
```
This feels like it could miss cases where we got a super clear indication a new test was starting (`=== RUN`) but misparsed the results line (maybe they change the duration format to nanoseconds or something). Seems like we should give the test data or test parser a chance to indicate it is complete, and close out the current test case, AND give the test parser a chance to indicate a new test is starting.

Initializing test and suite data with "unknown" name, result, duration, etc, populating those things as we know them, and pushing results into the current suite and moving on with a new test data if we see a new test starting makes us more robust to parse or output errors.

liggitt replied:
I'd also like to include *all* the lines relevant to the test (from `=== RUN` to the final results line) in the `system-out` JUnit field, to help quickly debug parse issues down the road. Hopefully that would simplify the parse as well.

liggitt replied:
The benefit of JUnit is the test case and test suite rollup, the summary info, and tracking run times and regressions over runs... it won't matter at all if two extra lines are in the `system-out` capture of every test.

php-coder commented on cmd/junitreport/junitreport.go:
```diff
@@ -0,0 +1,55 @@
+
+	err = options.Run()
+	if err != nil {
+		fmt.Printf("Error generating output: %v\n", err)
```
Just interesting: why we are not printing errors to stderr?

stevekuznetsov replied:
Good catch

soltysh commented on cmd/junitreport/pkg/api/string.go:
```diff
@@ -0,0 +1,35 @@
+func (t *TestCase) String() string {
+	var result string
+	var message string
+	var output string
```
Declare this in one line.

soltysh commented on cmd/junitreport/pkg/api/types.go:
```diff
@@ -0,0 +1,91 @@
+	XMLName xml.Name `xml:"testsuites"`
+
+	// Suites are the jUnit test suites held in this collection
+	Suites []*TestSuite
```
Is there any particular reason why you're doing arrays of pointers?

stevekuznetsov replied:
Pointers make child/parent relationships easier to mess with, but otherwise no.

stevekuznetsov replied:
Also `omitempty` needs a pointer to a struct to omit if empty.

stevekuznetsov commented:
@soltysh @liggitt I've addressed most comments. PTAL

The `gotest` parser is rewritten.
Unit tests were added for builders and the `gotest` parser. I will fix up the `stack` parser and the two parsers will be all we need to get `test-go`, `test-integration`, `test-extended`, and `test-end-to-end` parsed.

stevekuznetsov commented:
I've updated the stack parser. I think this is ready.

stevekuznetsov commented:
@soltysh @liggitt added integration test to validate XML output

stevekuznetsov commented:
Maybe in this PR or a follow-up I will add `junitreport summarize`, a subcommand that will pretty-print the non-successes for a test run. Potential output:
```
Go testing finished, <num> tests executed, <num> tests failed, <num> tests skipped
Test Failures:
In package <pkg>:
<testName>
<test output>
...
Test Skips:
...
```

kargakis commented on cmd/junitreport/pkg/api/types.go:
```diff
@@ -0,0 +1,98 @@
+	Name string `xml:"name,attr"`
+
+	// NumTests records the number of tests in the TestSuite
+	NumTests uint `xml:"tests,attr"`
```
Do we really need this field if we can just pass `len(TestCases)`?

stevekuznetsov replied:
This is required for serialization to XML.

kargakis commented on cmd/junitreport/pkg/builder/nested/test_suites_builder.go:
```diff
@@ -0,0 +1,184 @@
+			Children: []*api.TestSuite{child},
+		}
+
+		if err := b.addToParent(parent); err != nil {
```
`return b.addToParent(parent)`

kargakis commented on cmd/junitreport/pkg/builder/nested/test_suites_builder.go:
```diff
@@ -0,0 +1,184 @@
+	for _, suite := range suites {
+		if suite.Name == newSuite.Name || !strings.HasPrefix(newSuite.Name, suite.Name) {
+			// if we're considering the suite itself or another suite that is not a pure
+			// prefix of the new suite, we are not consiering a parent suite and therefore
```
typo: consiering

kargakis commented on cmd/junitreport/pkg/cmd/junitreport.go:
```diff
@@ -0,0 +1,115 @@
+type testSuitesBuilderType string
+
+const (
+	flatBuilderType   testSuitesBuilderType = "flat"
```
Since you follow conventions from the main repo, I would expect these constants under your api package.

kargakis commented:
@stevekuznetsov can you move this under /tools so we can start addressing https://github.com/openshift/origin/issues/6430?

stevekuznetsov commented:
@kargakis addressed your comments

Re: CLI switch/case types - I chose to stick the `suitesBuilderType` in the command line part of the code because it's not really part of the API for the tool, it is just used to build the parser structure. We do the same thing in other complicated CLI commands like `sync groups`.

stevekuznetsov commented:
Added `junitreport summarize` which decodes a jUnit XML file and gives us a nice summary:
```
$ cat tools/junitreport/test/gotest/reports/9.txt | junitreport summarize
Of 6 tests executed in 0.400s, 4 succeeded, 1 failed, and 1 was skipped.

In suite "package/name/nested", test case "TestOne" failed:
=== RUN   TestOne
--- FAIL: TestOne (0.02 seconds)
	file_test.go:11: Error message
	file_test.go:11: Longer
		error
		message.

In suite "package/name/nested", test case "TestTwo" was skipped:
=== RUN   TestTwo
--- SKIP: TestTwo (0.03 seconds)
	file_test.go:11: Skip message
PASS
```

stevekuznetsov commented:
Removed line anchors from regex so we can still correctly parse test results that didn't print correctly, i.e. `some other text--- PASS: TestName (0.00s)`. Added unit test to all regex tests for this.

soltysh commented on tools/junitreport/README.md:
```diff
@@ -0,0 +1,47 @@
+
+## Installation
+
+In order to build and install `junitreport`, from the root of the OpenShift Origin repository, run `hack/buil-go.sh cmd/junitreport`. 
```
s/buil-go/build-go

soltysh replied:
tools/junitreport

soltysh commented on tools/junitreport/README.md:
```diff
@@ -0,0 +1,47 @@
+
+Currently, `junitreport` does not support the parsing of parallel test output.
+
+### Examples
```
I'd rather see examples applicable to origin resources, iow. against `hack/test-go.sh` rather than `go test`, the latter won't work because you don't have the env set properly (at least that was my error).

stevekuznetsov replied:
:+1: I was waiting for this pull to merge before I made changes to `hack/test-go.sh` as I had a `hack/test-go.sh` pull open before but it took too much maintenance with this not being in.

I envision a "reporting" flag for `hack/test-go.sh` that will run a different script almost, culminating in a `go test -v | junitreport > report.xml | tee junitreport summarize` instead of the current flow. The impetus behind `summarize` was that it seemed like a fool's game to try to make a regex nice (read: crazy) enough to do live-filtering of the output of `test go -v` to get the terminal output to not change when using this tool.

soltysh commented on tools/junitreport/pkg/parser/gotest/data_parser.go:
```diff
@@ -0,0 +1,142 @@
+		}, true
+	}
+	return map[string]string{}, false
+}
```
This function should one additional information, the overall time of a test suite. See my below output:
```
$ hack/test-go.sh pkg/build/controller/ -v -cover
=== RUN   TestHandleBuildConfig
--- PASS: TestHandleBuildConfig (0.00s)
=== RUN   TestHandleBuild
--- PASS: TestHandleBuild (0.00s)
=== RUN   TestHandlePod
--- PASS: TestHandlePod (0.00s)
=== RUN   TestCancelBuild
--- PASS: TestCancelBuild (0.00s)
...
--- PASS: TestNewImageIDNoDockerRepo (0.00s)
PASS
coverage: 86.3% of statements
ok  	github.com/openshift/origin/pkg/build/controller	0.010s
hack/test-go.sh took 3 seconds
```

As you see from above output each tests takes so little that it never shows anything other than `0.00s`, but the overall test suite time is significant enough `0.010s` that it's printed. Since you're summing each test time as a summary you'll showing a bit different information than `go test` does. This can be easily verified if you modify your test case file `tools/junitreport/pkg/parser/gotest/test/1.txt` like above and you'll get the problem I'm talking about.

stevekuznetsov replied:
I think this is working as intended at the moment. I couldn't find a way to reconcile the time at  the end of the suite and the times in the test cases. Specifically, when we are creating nested test suites, I don't really know a good way to update the tree with correct timing if we have two different sources of truth for the timing. If we can come up with a solid update strategy I think we can parse this field as well.

stevekuznetsov replied:
I'm also working on a change upstream to allow for more precise time output on tests... but that may take a while.

stevekuznetsov replied:
Duration isn't a property, but a first-class field on the package. This will take a larger refactor. I don't think my interfaces are worth it - there are too many differences between `go test` output and others, adding an `ExtractDuration` method to the package data extractor would not be a good idea - output of `test/cmd` for instance won't use it.

soltysh commented on tools/junitreport/junitreport.go:
```diff
@@ -0,0 +1,134 @@
+  $ go test -v -cover ./... | junitreport --suites=nested --roots=github.com/maintainer > report.xml
+
+  # Describe failures and skipped tests in an existing jUnit XML file
+  $ cat report.xml | %[1]s summarize
```
ditto

stevekuznetsov replied:
Does this comment mean you want to see the usage with `hack/test-go` ?

soltysh commented on tools/junitreport/pkg/parser/gotest/parser.go:
```diff
@@ -0,0 +1,100 @@
+			}
+			currentTest = &api.TestCase{}
+			currentTestResult = api.TestResultFail
+			currentTestOutput = []string{}
```
Is it intended that if my passing test spits some output eg.:
```
$ hack/test-go.sh pkg/build/controller/ -v -cover
=== RUN   TestHandleBuildConfig
E0105 11:11:52.973414    4347 config_controller.go:66] error instantiating Build from BuildConfig /testBuildConfig: error
--- PASS: TestHandleBuildConfig (0.00s)
...
```
You're ignore this, and only skipped and failed output is part of the report?

stevekuznetsov replied:
This is how jUnit XML is formatted, actually. The output (and/or message) are part of the XML nodes that indicate failure or skippage. jUnit doesn't support capturing the output for passing tests. Since we're going to be saving full verbosity test output as an artifact on Jenkins I think this is OK.

stevekuznetsov commented:
@soltysh Updated parser to over-ride the Gotest package duration from the package declaration line.

stevekuznetsov commented:
@soltysh Updated to remove redundant test data, added `test/integration.sh` invocation to `hack/test-cmd.sh`. @deads2k let's think of a better place to put that test when you're done with meetings.

stevekuznetsov commented:
[test]

stevekuznetsov commented:
@soltysh @deads2k created `hack/test-tools.sh` and added it to target `make test-int-plus`. It expects a test at `tools/<tool>/test/integration.sh` and if one exists, it is run.

soltysh commented:
LGTM, @deads2k if you don't have any objects I'd like to merge this today.

stevekuznetsov commented on tools/junitreport/.gitignore:
```diff
@@ -0,0 +1 @@
+junitreport
```
@soltysh why doesn't this correctly pick up the binary `junitreport` when I run `go build ./...` from this directory?

soltysh commented:
[merge]

soltysh commented:
This PR does not modify origin itself, so it's a flake.
re-[merge]
